### PR TITLE
ui: Upgrade React version and react-dependent libraries

### PR DESCRIPTION
**Pull request by koorosh**
- Has to merged after: https://github.com/cockroachdb/cockroach/pull/43685
- Depends on https://github.com/cockroachdb/yarn-vendored/pull/13

Current PR has affected a lot of files but most of changes are almost the same
and do not affect business logic.

With current changes, react is upgraded to latest version (16.12.0) and
following libraries are upgraded as well: react-redux, redux, redux-saga.
These upgrades of major versions has required to do some code adjustments
to properly migrate to newest versions.
- redux. Provide valid type arguments for Dispatch type.
- redux-saga. Changed import and usage of 'delay' function.
- react-redux. `connect` function failed to correctly resolve returned type
for `mapDispatchToProps` argument. As result it is provided as a function
instead of object argument. This hack allows implicitly resolve types.
- `CachedDataReducer` class is extended with one more optional type argument
to specify the list of allowed `actionNamespace` literals instead of `string`.
It was necessary to have strictly defined CachedDataReducer interface.
- `normalizeConnectedComponent` function is used to do simple mapping from
react's ExoticComponent to valid react component. This resolves issue of
providing ConnectedComponents (which are ExoticComponent's, not regular react
components) to Route component which has strict type validation of component types.

### CODE REVIEW

File: pkg/ui/ccl/src/views/shared/containers/licenseSwap/index.tsx [hunk 1]
```diff
@@ -53,11 +53,11 @@ export default function swapByLicense<TProps>(
   OSSComponent: React.ComponentClass<TProps>,
   CCLComponent: React.ComponentClass<TProps>,
   // tslint:enable:variable-name
-): React.ComponentClass<TProps> {
+) {
   const ossName = getComponentName(OSSComponent);
   const cclName = getComponentName(CCLComponent);
 
-  class LicenseSwap extends React.Component<TProps & OwnProps> {
+  class LicenseSwap extends React.Component<TProps & OwnProps & any> {
```
Comment (Koorosh) [inline pkg/ui/ccl/src/views/shared/containers/licenseSwap/index.tsx]: Following hack with **&any** type is required to make changes related to types only and do not rewrite component logic.

**Review by dhartunian — CHANGES_REQUESTED** [review]
Thanks for this huge undertaking @Koorosh! My only major concern is whether the restructuring the `mapDispatchToProps` is necessary. See my comments for the approach I looked at.

File: pkg/ui/src/views/cluster/containers/events/index.tsx [hunk 2]
```diff
@@ -178,9 +176,9 @@ const eventBoxConnected = connect(
       eventsValid: eventsValidSelector(state),
     };
   },
-  {
+  () => ({
```
[the remark this answers is not in the document]
Comment (Koorosh) [reply pkg/ui/src/views/cluster/containers/events/index.tsx]: Agree. Removed unnecessary function wrappers across all project

File: pkg/ui/src/redux/cachedDataReducer.ts [hunk 3]
```diff
@@ -180,7 +180,7 @@ export class CachedDataReducer<TRequest, TResponseMessage> {
    *   state given the global state object
    */
   refresh = <S>(req?: TRequest, stateAccessor = (state: any, _req: TRequest) => state.cachedData[this.actionNamespace]) => {
-    return (dispatch: Dispatch<S>, getState: () => any) => {
+    return (dispatch: Dispatch<Action, TResponseMessage>, getState: () => S) => {
```
[the remark this answers is not in the document]
Comment (Koorosh) [reply pkg/ui/src/redux/cachedDataReducer.ts]: Awesome, it makes code much cleaner! Thanks!

**Review by dhartunian — APPROVED** [review]
👍 LGTM. Thanks for the updates!

**Pull request merged**